translator/trace/oc_to_otlp.go

```diff
@@ -0,0 +1,420 @@
+const (
+	ocAttributeProcessStartTime  = "opencensus.starttime"
+	ocAttributeProcessID         = "opencensus.pid"
+	ocAttributeExporterVersion   = "opencensus.exporterversion"
+	ocAttributeResourceType      = "opencensus.resourcetype"
+	ocTimeEventMessageEventType  = "opencensus.timeevent.messageevent.type"
+	ocTimeEventMessageEventID    = "opencensus.timeevent.messageevent.id"
+	ocTimeEventMessageEventUSize = "opencensus.timeevent.messageevent.usize"
+	ocTimeEventMessageEventCSize = "opencensus.timeevent.messageevent.csize"
+)
+
+func ocToOtlp(td consumerdata.TraceData) consumerdata.OTLPTraceData {
+
+	if td.Node == nil && td.Resource == nil && len(td.Spans) == 0 {
+		return consumerdata.OTLPTraceData{}
+	}
+
+	resource := ocNodeResourceToOtlp(td.Node, td.Resource)
+
+	resourceSpans := &otlptrace.ResourceSpans{
+		Resource: resource,
+	}
+	resourceSpanList := []*otlptrace.ResourceSpans{resourceSpans}
+
+	if len(td.Spans) != 0 {
+		resourceSpans.Spans = make([]*otlptrace.Span, 0, len(td.Spans))
+
+		for _, ocSpan := range td.Spans {
+			if ocSpan == nil {
+				// Skip nil spans.
+				continue
+			}
+
+			otlpSpan := ocSpanToOtlp(ocSpan)
+
+			if ocSpan.Resource != nil {
+				// Add a separate ResourceSpans item just for this span since it
+				// has a different Resource.
+				separateRS := &otlptrace.ResourceSpans{
+					Resource: ocNodeResourceToOtlp(nil, ocSpan.Resource),
+					Spans:    []*otlptrace.Span{otlpSpan},
+				}
+				resourceSpanList = append(resourceSpanList, separateRS)
+			} else {
+				// Otherwise add the span to the first ResourceSpans item.
+				resourceSpans.Spans = append(resourceSpans.Spans, otlpSpan)
+			}
+		}
+	}
+
+	return consumerdata.NewOTLPTraceData(resourceSpanList)
+}
+
+func timestampToUnixnano(ts *timestamp.Timestamp) uint64 {
+	return uint64(internal.TimestampToTime(ts).UnixNano())
+}
+
+func ocSpanToOtlp(ocSpan *octrace.Span) *otlptrace.Span {
+	attrs, droppedAttrCount := ocAttrsToOtlp(ocSpan.Attributes)
+	events, droppedEventCount := ocEventsToOtlp(ocSpan.TimeEvents)
+	links, droppedLinkCount := ocLinksToOtlp(ocSpan.Links)
+
+	childSpanCount := int32(0)
+	if ocSpan.ChildSpanCount != nil {
+		childSpanCount = int32(ocSpan.ChildSpanCount.Value)
+	}
+
+	otlpSpan := &otlptrace.Span{
+		TraceId:                ocSpan.TraceId,
+		SpanId:                 ocSpan.SpanId,
+		Tracestate:             ocTraceStateToOtlp(ocSpan.Tracestate),
+		ParentSpanId:           ocSpan.ParentSpanId,
+		Name:                   truncableStringToStr(ocSpan.Name),
+		Kind:                   ocSpanKindToOtlp(ocSpan.Kind, ocSpan.Attributes),
+		StartTimeUnixnano:      timestampToUnixnano(ocSpan.StartTime),
+		EndTimeUnixnano:        timestampToUnixnano(ocSpan.EndTime),
+		Attributes:             attrs,
+		DroppedAttributesCount: droppedAttrCount,
+		Events:                 events,
+		DroppedEventsCount:     droppedEventCount,
+		Links:                  links,
+		DroppedLinksCount:      droppedLinkCount,
+		Status:                 nil,
```

Comment:
Is this intentionally nil?

Reply:
No, thanks for catching! Completely missed the Status translation.

Reply:
Added, please have a look.